src/ngraph/pass/reshape_sinking.cpp

```diff
@@ -586,6 +626,14 @@ bool ngraph::pass::ReshapeSinking::run_on_function(shared_ptr<ngraph::Function>
         {
             sink_concat(concat, reorders, reshapes_to_delete);
         }
+        else if (auto reduce = as_type_ptr<op::Max>(n))
```

Comment:
Can we use the same logic and handle all reductions (Arithmetic, Logical, Index)?

Reply:
i think so as they all reduce axis similarly, we can enable it. This PR only enables reducing on one axis, as i havent fullly tested reducing on multiple axis. 
Also, v1 ops that have keep_dims  need to be handled slightly differently.

Reply:
Why are you using a template? Using `shared_ptr<LogicalReduction>` should work.
It looks like `IndexReduction` and `LogicalReduction` are suffering from opset1 and dynamic shape change collisions and need to be made consistent.

Reply:
inside this function we need to know the actual op type as we create a new replacement op. Using shared_ptr<LogicalReduction> works to keep things common, but we will have to identify op type for creation. I looked at `copy_with_new_args` but that doesnt work seamlessly either. Template allowed to pass type for op (re)creation.

Reply:
You could do this with `copy_with_new_inputs` or the node factory and then set the attributes , but that is better done with a separate major refactoring PR.